From 31a9b2d84a37299bf2cd499448ead223d1d95bc6 Mon Sep 17 00:00:00 2001
From: Erhan Bilgili <erhan.bilgili@gmail.com>
Date: Wed, 29 Jul 2026 09:34:02 +0200
Subject: [PATCH] win32u, winewayland: Fix maximized frameless window geometry.

Preserve the border when detecting collapsed captions and recognize
captionless maximized resize frames as non-client geometry.

Avoid expanding frameless configure sizes twice.

(cherry picked from commit 31a9b2d84a37299bf2cd499448ead223d1d95bc6)

Source: https://github.com/nanomatters/wine-salkim/commit/31a9b2d84a37299bf2cd499448ead223d1d95bc6
---
 dlls/win32u/window.c          | 5 ++++-
 dlls/winewayland.drv/window.c | 8 --------
 2 files changed, 4 insertions(+), 9 deletions(-)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index 3ab847ede89..bef8841b245 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -2100,7 +2100,7 @@ static BOOL has_collapsed_caption( UINT style, UINT ex_style,
     if (top < 0 || client->left < window->left ||
         client->right > window->right || client->bottom > window->bottom)
         return FALSE;
-    if (!adjust_window_rect( &border, style & ~WS_CAPTION, FALSE, ex_style, get_system_dpi() ))
+    if (!adjust_window_rect( &border, style & ~WS_DLGFRAME, FALSE, ex_style, get_system_dpi() ))
         return FALSE;
 
     border_top = max( 0, -border.top );
@@ -2111,6 +2111,9 @@ static BOOL is_frameless_window( HWND hwnd, UINT style, UINT ex_style,
     const struct window_rects *rects )
 {
     if (!window_has_frame_style( style )) return FALSE;
+    if ((style & (WS_MAXIMIZE | WS_THICKFRAME)) == (WS_MAXIMIZE | WS_THICKFRAME) &&
+        (style & WS_CAPTION) != WS_CAPTION)
+        return TRUE;
     return has_collapsed_frame( &rects->window, &rects->client, get_custom_frame( hwnd ) ) ||
            has_collapsed_caption( style, ex_style, &rects->window, &rects->client );
 }
diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index 45c3b683b65..9d6baf83739 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -1026,14 +1026,6 @@ static void wayland_configure_window(HWND hwnd)
         window_height += offset_y;
         window_width += offset_x;
     }
-    else if (data->frameless)
-    {
-        window_width += (data->rects.client.left - data->rects.window.left) +
-                        (data->rects.window.right - data->rects.client.right);
-        window_height += (data->rects.client.top - data->rects.window.top) +
-                         (data->rects.window.bottom - data->rects.client.bottom);
-    }
-
     SetRect(&rect, 0, 0, window_width, window_height);
     OffsetRect(&rect, data->rects.window.left, data->rects.window.top);
     if (!IsRectEmpty(&rect)) rect = window_rect_from_visible(&data->rects, rect);
-- 
2.55.0

